File: rust/src/ui/skills_view.rs
// Read-only skills view (SPEC 21.3): top line "N skills" + rescan hint,
// groups by source (case-insensitive name sort within group, done in
// skills.rs::scan), scrollable. Zero background cost: scan once on first
// open, cached in AppState; 'r' forces a rescan (SPEC 21.2).

use crate::app::App;
use crate::theme::Palette;
use ratatui::layout::{Constraint, Layout, Rect};
use ratatui::style::{Modifier, Style};
use ratatui::text::{Line, Span};
use ratatui::widgets::{Block, Padding, Paragraph};
use ratatui::Frame;

const FOOTER: &str = "↑/↓ Scroll · r Rescan · Esc Back · q Quit";

#[derive(Clone)]
pub enum Row {
    Group(String),
    Item { name: String, description: String },
}

pub fn draw(f: &mut Frame, app: &App, p: &Palette) {
    let chunks = Layout::vertical([
        Constraint::Length(3),
        Constraint::Min(3),
        Constraint::Length(1),
    ])
    .split(f.area());

    let n_items = app
        .skills_rows
        .iter()
        .filter(|r| matches!(r, Row::Item { .. }))
        .count();
    let summary = if app.skills_loading {
        "Scanning...".to_string()
    } else {
        format!("{n_items} skills")
    };
    let title = Line::from(vec![
        Span::styled(
            "Kimi Skills",
            Style::default().fg(p.text_primary).add_modifier(Modifier::BOLD),
        ),
        Span::styled(format!("   {summary}"), Style::default().fg(p.text_secondary)),
    ]);
    f.render_widget(
        Paragraph::new(title).block(Block::default().padding(Padding::new(2, 0, 1, 0))),
        chunks[0],
    );

    let inner = Rect {
        x: chunks[1].x,
        y: chunks[1].y + 1,
        width: chunks[1].width,
        height: chunks[1].height.saturating_sub(1),
    };

    // Build one display line per row; items get a name line + description line.
    let mut lines: Vec<Line> = Vec::new();
    // Map: display line index of each skills_rows entry (first line of it)
    let mut row_line: Vec<usize> = Vec::new();
    for (idx, row) in app.skills_rows.iter().enumerate() {
        row_line.push(lines.len());
        match row {
            Row::Group(source) => {
                if idx > 0 {
                    lines.push(Line::raw(""));
                }
                lines.push(Line::from(Span::styled(
                    source.clone(),
                    Style::default().fg(p.accent).add_modifier(Modifier::BOLD),
                )));
            }
            Row::Item { name, description } => {
                let selected = idx == app.skills_sel;
                let name_style = if selected {
                    Style::default()
                        .fg(p.text_primary)
                        .bg(p.button_hover)
                        .add_modifier(Modifier::BOLD)
                } else {
                    Style::default().fg(p.text_primary).add_modifier(Modifier::BOLD)
                };
                let desc_style = if selected {
                    Style::default().fg(p.text_secondary).bg(p.button_hover)
                } else {
                    Style::default().fg(p.text_secondary)
                };
                lines.push(Line::from(Span::styled(format!("  {name}"), name_style)));
                lines.push(Line::from(Span::styled(format!("    {description}"), desc_style)));
            }
        }
    }

    // Keep the selected row's first line inside the viewport
    let viewport = inner.height as usize;
    let sel_line = row_line.get(app.skills_sel).copied().unwrap_or(0);
    let total = lines.len();
    let max_scroll = total.saturating_sub(viewport);
    let scroll = if sel_line >= viewport {
        (sel_line + 1).saturating_sub(viewport).min(max_scroll)
    } else {
        0
    };

    f.render_widget(
        Paragraph::new(lines).scroll((scroll as u16, 0)),
        Rect { ..inner },
    );

    f.render_widget(
        Paragraph::new(FOOTER).style(Style::default().fg(p.text_secondary)),
        chunks[2],
    );
}

File: go/internal/tui/skillsview.go
// Read-only skills view (SPEC 21.3): "N skills" summary + rescan hint, rows
// grouped by source (case-insensitive name sort happens in core.ScanSkills),
// scrollable with the selected item kept in the viewport. Zero background
// cost: scan once on first open, cached in AppState; 'r' forces a rescan
// (SPEC 21.2). Layout translated from rust/src/ui/skills_view.rs.
package tui

import (
	"fmt"

	"github.com/shawn-0106t/kimi-planbar-tui/go/internal/core"
)

const skillsFooter = "↑/↓ Scroll · r Rescan · Esc Back · q Quit"

// SkillsRow is a flattened display row: a group header or a skill item.
type SkillsRow struct {
	IsGroup     bool
	Source      string // group header text
	Name        string // item only
	Description string // item only
}

// skillsView returns the content rows above the footer: blank, padded title,
// blank (3-row title block), blank (inner-rect top offset), then the visible
// portion of the list.
func skillsView(m *Model, p core.Palette) []Line {
	title := styledSpan("Kimi Skills", p.TextPrimary)
	title.Bold = true
	summary := fmt.Sprintf("%d skills", m.countSkillItems())
	if m.skillsLoading {
		summary = "Scanning..."
	}
	summarySpan := styledSpan("   "+summary, p.TextSecondary)

	rows := []Line{
		blankOf(p),
		{Spans: []Span{title, summarySpan}},
		blankOf(p),
		blankOf(p),
	}

	// Build one or two display lines per row; remember each row's first line.
	lines := []Line{}
	rowLine := make([]int, len(m.skillsRows))
	for idx, row := range m.skillsRows {
		rowLine[idx] = len(lines)
		switch {
		case row.IsGroup:
			if idx > 0 {
				lines = append(lines, blankOf(p))
			}
			s := styledSpan(row.Source, p.Accent)
			s.Bold = true
			lines = append(lines, Line{Spans: []Span{s}})
		default:
			name := styledSpan("  "+row.Name, p.TextPrimary)
			name.Bold = true
			desc := styledSpan("    "+row.Description, p.TextSecondary)
			if idx == m.skillsSel {
				name.Bg = p.ButtonHover
				desc.Bg = p.ButtonHover
			}
			lines = append(lines, Line{Spans: []Span{name}}, Line{Spans: []Span{desc}})
		}
	}

	// Keep the selected row's first line inside the viewport.
	viewport := m.height - 5 // 3 title rows + inner offset + footer
	if viewport < 0 {
		viewport = 0
	}
	selLine := 0
	if m.skillsSel >= 0 && m.skillsSel < len(rowLine) {
		selLine = rowLine[m.skillsSel]
	}
	total := len(lines)
	maxScroll := total - viewport
	if maxScroll < 0 {
		maxScroll = 0
	}
	scroll := 0
	if selLine >= viewport {
		scroll = selLine + 1 - viewport
		if scroll > maxScroll {
			scroll = maxScroll
		}
	}
	if scroll < 0 {
		scroll = 0
	}
	end := scroll + viewport
	if end > total {
		end = total
	}
	if scroll < end {
		rows = append(rows, lines[scroll:end]...)
	}
	return rows
}

// skillsFooterRow builds the pinned footer.
func skillsFooterRow(p core.Palette) Line {
	return Line{Spans: []Span{styledSpan(skillsFooter, p.TextSecondary)}}
}

// firstItem is the index of the first selectable row (group headers are not
// selectable); -1 when there is none.
func firstItem(rows []SkillsRow) int {
	for i, r := range rows {
		if !r.IsGroup {
			return i
		}
	}
	return -1
}

// moveSkillsSel moves the highlight to the next/previous item row, skipping
// group headers and wrapping (rust app.rs move_skills_sel).
func moveSkillsSel(rows []SkillsRow, sel, dir int) int {
	if len(rows) == 0 {
		return 0
	}
	i := sel
	for range len(rows) {
		i = ((i+dir)%len(rows) + len(rows)) % len(rows)
		if !rows[i].IsGroup {
			return i
		}
	}
	return sel
}

// countSkillItems counts selectable rows for the summary line.
func (m *Model) countSkillItems() int {
	n := 0
	for _, r := range m.skillsRows {
		if !r.IsGroup {
			n++
		}
	}
	return n
}

// flattenSkills turns a scan result into display rows (rust App::set_skills).
func flattenSkills(skills []core.SkillInfo) []SkillsRow {
	rows := []SkillsRow{}
	lastSource := ""
	for i, s := range skills {
		if i == 0 || lastSource != s.Source {
			rows = append(rows, SkillsRow{IsGroup: true, Source: s.Source})
		}
		lastSource = s.Source
		rows = append(rows, SkillsRow{Name: s.Name, Description: s.Description})
	}
	return rows
}
